### ISSUE TITLE
Code refactoring

### ISSUE THREAD

**Issue by mahaalbashir** [AI-SDC/ACRO#195]
Refactoring the function create_crosstab_masks in acro_table.

Looking at the levels of indentation and the numbers of lines in this method near where you calculate threshold masks etc., (lines 742 onwards) , I wonder if it is time to refactor some of this code into a series of functions called something like get_mask_for_aggfuncc()

**Comment by ssrhaso** [comment]
@jim-smith  Could you have a look at this and let me know steps moving forward

**Comment by jim-smith** [comment]
@ssrhaso 
(cc @JessUWE ) this really helpful.
sorry for the radio silence -been on leave.
Hoping to spend some time this afternoon working in a parallel branch where the logic is driven by a set of rules. 
So maybe worth waiting a few days while I work out how that might affect things.

One big thing to think about is whether we should be 
- *removing* rows/columns where all the cells are empty, or
-  marking them as *Not a Number*  - then possibly later replacing NaN with a string f'< {THRESHOLD} ' 
I think we should probably consult about that because:
-  we might be inadvertently revealing information in some edge cases
- it introduces a whole world of pain and complexity - which @JessUWE has just been dealing with wrt how it shows up in the viewer.

**Comment by ssrhaso** [comment]
> [@ssrhaso](https://github.com/ssrhaso) (cc [@JessUWE](https://github.com/JessUWE) ) this really helpful. sorry for the radio silence -been on leave. Hoping to spend some time this afternoon working in a parallel branch where the logic is driven by a set of rules. So maybe worth waiting a few days while I work out how that might affect things.
> 
> One big thing to think about is whether we should be
> 
> * _removing_ rows/columns where all the cells are empty, or
> * marking them as _Not a Number_  - then possibly later replacing NaN with a string f'< {THRESHOLD} '
>   I think we should probably consult about that because:
> * we might be inadvertently revealing information in some edge cases
> * it introduces a whole world of pain and complexity - which [@JessUWE](https://github.com/JessUWE) has just been dealing with wrt how it shows up in the viewer.

No problem, please keep me updated - I am thoroughly enjoying all the world on SACRO-ML in particular and will be shortly opening a PR for the QMIA method with Shamy

**Comment by JessUWE** [comment]
Hassan is still waiting on Jim @jim-smith for specific direction has @jim-smith has said he would get back to @ssrhaso

**Comment by jim-smith** [comment]
@ssrhaso if you look a  the pushes in the branch 350 -need json files and two new files (checks,py, aggregations.pf) and then the edits in acro_regressions completed)  and the crosstab (not finalised)  function (but not the rest of acro_tables) you may get a sense of where I am heading with this.

I've got some more edits to complete and push then will put in a simplified diagram. I hope to then hand it over to you guys to tidy up and do the edits for pivot_table and all the plots.


Biggest changes are that:

- in regressions you'll see I've pulled out the hard-coded calls to checks. Instead there  are just t wo lines in each method that do all the sdc stuff, all the rest is jsut calling standard stats models functions. Those two are:
   - declare the name of the type of analysis
   - pass the model and that name into a function `run_checks_for_analysis(name, model)` 
- in crosstab  we always set `dropna=False` so we can lose lots of code that deals with  table structure. It all works fine except for when  we apply suppression in the presence of marginal totals.
    -- what I'm going to do there is make a deep copy of the data then create and queries that remove all the  data lying in 'sensitive' cells then just make a new call to the crosstab function with the redacted dataset
- but again in cross tab you jsut make a list of the aggregation functions the researcher wants to report on, then loop through those with the same two lines of code as in the regression funcs

**Comment by jim-smith** [comment]
**really useful piece of work bit now superseded by #350
